#### Lesson 5. Date: 26.10.23

**Topics**:
> 1. Time/Datetime
> 2. Zip
> 3. Namespace (global, local, nonclocal)
> 4. Function (Parameters: Args/Kwargs, Lambda) 
> 5. Try-Except - work with error
> 6. Map/Reduce/Filter


**Materials**:
> 1. [Time/Datetime](https://docs.python.org/3/library/datetime.html)
> 2. [Namespace](https://realpython.com/python-namespaces-scope/)
> 3. [Try/Except](https://www.geeksforgeeks.org/try-except-else-and-finally-in-python/) 
> 4.  [Map/Reduce/Filter](https://book.pythontips.com/en/latest/map_filter.html)

In [1]:
# Time - для роботи із часом
import time # імпорт додаткового функціоналу для робити з часом

current_time = time.time()  # Повертає кількість секунд, що пройшли з 1 січня 1970 року
print(current_time)

1698338767.938303


In [4]:
print("Початок")
# -> API 
for i in range(0, 10):
    # -> API
    time.sleep(2)  # Затримка на 3 секунди
# -> API
print("Кінець")

Початок
Кінець


In [5]:
timestamp = time.time()  # Поточний час
local_time = time.localtime(timestamp)  # Перетворення часу на локальний часовий пояс
print(local_time, type(local_time))
print(local_time.tm_year, local_time.tm_mday, local_time.tm_mon) # отримати складову часу за допомогою дот нотації 

time.struct_time(tm_year=2023, tm_mon=10, tm_mday=26, tm_hour=19, tm_min=52, tm_sec=7, tm_wday=3, tm_yday=299, tm_isdst=1) <class 'time.struct_time'>
2023 26 10


In [24]:
# Datetime - бібліотека для зручних математичних обрахунків із датою
from datetime import datetime, date  # НЕ рекомендую

date = date(2023, 10 , 26)  # Створення об'єкта дати зі значеннями "2023-10-26"
print(date, type(date))

now = datetime.now()  # Повертає поточний об'єкт дати та часу
print(now)


2023-10-26 <class 'datetime.date'>
2023-10-26 19:58:16.163699


In [33]:
import datetime

date1 = datetime.date(2023, 4, 10)
date2 = datetime.date(2024, 4, 10)

delta = date2 - date1  # Різниця між датами
print(delta.days, delta)  # Виведення кількості днів між датами

366 366 days, 0:00:00


In [48]:
### Zip - Функція яка дозволяє поєднати два об'єкти в один
list1 = [1, 2, 3, 4]
list2 = ['a', 'b', 'c', 'd']

# zipped = zip(list1, list2)  # Злиття списків
# print(zipped, list(zipped))  # Виведення злитого ітерабельного об'єкту у вигляді списку

<zip object at 0x10dead500> [(1, 'a'), (2, 'b'), (3, 'c'), (4, 'd')]


In [59]:
zipped = tuple(zip(list1, list2))
print(zipped[2][0])

3


In [87]:
arr = ['1', '2', '3', '4', '5']
int_arr = map(lambda x: int(x), arr)
print(list(int_arr))   

[1, 2, 3, 4, 5]


In [60]:
string_1 = 'str'
string_2 = 'tmp'
print(list(zip(string_1,string_2)))

[('s', 't'), ('t', 'm'), ('r', 'p')]


In [61]:
dict_ = {1:'a', 2:'b', 3:'c'}
tuple_1 = (4,5,6)

print(list(zip(dict_.values(),tuple_1)))

[('a', 4), ('b', 5), ('c', 6)]


In [62]:
list1 = [1, 2, 3]
list2 = ['a', 'b', 'c']
list3 = ['A', 'B', 'C']

zipped = zip(list1, list2, list3)  # Злиття списків
print(zipped, list(zipped)) 

<zip object at 0x10dea10c0> [(1, 'a', 'A'), (2, 'b', 'B'), (3, 'c', 'C')]


#### Namespace - простір імен, по якому Python знаходить змінні
Ключові слова:
> **global** - змінна буде належати до глобальної області видимості \
> **local** - для роботи із локальними змінними \
> **nonlocal** - роботи із змінною на рівень вище   

**LEGB RULE** - Local-Enclo-Global-Built-in

In [63]:
count = 0 # global 

def increment_value(count): # local var
    count += 1
    return count

print(increment_value(count))
print(count)

1
0


In [71]:
from math import pi

#PI = 3.14
pi = 3.14 # global
def print_pi():
    print(pi)
    
print_pi()

3.14


In [ ]:
def print_count():
    print(count)
    
print_count()

In [73]:
count = 0 # global 

def increment_value():
    # local 
    global count # global var
    count += 1
    
print(count)
increment_value()
#increment_value()

print(count)


0
1


In [ ]:
def add(a, b):
    return a + b 

def even(number):
    return number % 2 == 0

result = add(5, 5)

even(result)

In [ ]:
count = 0
def increment_value(count: int): 
    count += 1
    return count 
print(count)

count = increment_value(count)
print(count)

In [ ]:
i = 0 
while i < 15:
    a = 5 
    i += a 
    
print(i, a)

In [2]:
def example(i: int = 10):
    def example1(i: int = 20):
        print(i)
        
    print(i) # Local
    example1() # Local
    example1() # Local
    
    
example()

10
20
20


In [4]:
def outer():
    x = "local"

    def inner():
        # x = x + 'local'
        nonlocal x
       # x = "nonlocal" # Змінна бути у зовнішій функції 
        y = x + 'local'
        print(x + 'local', y)
        return x 
    
    
    inner() #  виклик внутріш
    
    print("Після виклику inner():", x)

outer()


locallocal locallocal
Після виклику inner(): local


#### Function - іменнована область памяті, яка зберігає набір інструкція, які можна повторно використовувати
**Args** - зірочка(*) вказує що може бути необмежена кількість аргументів функції \
**Kwargs** - дві зірочки(**) вказує що функція може працювати та приймати значення у форматі key-value

In [ ]:
# Simple function 
def add(a: int, b : int) -> int:
    return a + b 

print(add(a = 5, b = 5))

In [ ]:
def my_function(*args): # aster, довільна кількість аргументів функції 
    print(type(args))
    for arg in args: # проходження по аргументах
        print(arg)

my_function(1, 2, 3, 4, 5)  # Виведення позиційних аргументів


In [ ]:
def sum_numbers(*args):
    return sum(args)

numbers = [i for i in range(0, 15)]
result = sum_numbers(*numbers)  # Розпаковка списку numbers у вигляді аргументів
print(result)  # Виведення суми чисел


In [ ]:
def my_function(**kwargs): # прийняти позіційні та ключові аргументи 
    print(type(kwargs))
    for key, value in kwargs.items(): 
        print(key, value)

my_function(a=1, b=2, c=3)  # Виведення ключових аргументів


#### Task 1
Візьміть як вхідні дані два рядки: список імен студентів, розділених символом “,”, і список їхніх кінцевих оцінок (цілі числа), розділених “,”.
Надрукуйте імена всіх учнів, чий підсумковий бал був на рівні мінімум 90. Для цієї вправи використовуйте якомога менше рядків коду

In [ ]:
names = 'Sasha,Alex,Yaroslav,Masha,Petro'.split(',')
marks_math = '90,60,96,75,59'.split(',')
marks_biolagy = '90,60,96,75,59'[::-1].split(',')

result = {}
for i in zip(names, marks_math, marks_biolagy):
    result[i[0]] = (int(i[1]), int(i[2]))
print(result)

for key, value in result.items():
    if value[1] >= 90:
         print(key, value[1])

In [ ]:
mark = []
for i in '90,60,96,75,59'.split(','):
    mark.append(int(i))
    
print(mark)

In [ ]:
def task1(**kwargs):
    print(kwargs)
    result = {}
    for key, value in kwargs.items():
        print(key, value)
        if value >= 90:
            result[key] = value
    return result
    

names = 'Sasha,Alex,Yaroslav,Masha,Petro'
marks = '90,60,96,75,59'
names = names.split(',')
marks = [int(i) for i in marks.split(',')]


for i in zip(names, marks):
    print(i, i[0], i[1])

#print(task1(**{i[0]: i[1] for i in zip(names, marks)}))


In [ ]:
print(task1(**{i[0]: i[1] for i in zip(names, marks)}))

In [ ]:
def my_function(a:int, b, c, d = 5, *args, **kwargs): # прийняти позіційні та ключові аргументи 
    
    def work_with_args():
        return sum(i for i in args)
    
    def work_with_kwargs():
        for key, value in kwargs.items(): 
            print(key, value)
        
    print(f'A = {a} B = {b} C = {c} D = {d}\nArgs = {args}\nKwargs = {kwargs} ')
    # print(work_with_kwargs())
    return a + b + c + d + work_with_args()

my_function(1,2,3,4,5,6,7,8,9,10,11,e=1, f=2, g=3)  #

### Важлива пріоритетність передач: 
1. Спочатку обов'язкові аргументи
2. Дефолтні аргументи 
3. Args
4. Kwargs


#### Lambda - функція, анонімна функція
add = lambda x, y: x + y \
**lambda** - key-word \
x, y - arguments \
x + y - body


In [80]:
add = lambda x, y : x + y # def _ (param1, param2): дія 
div = lambda x,y : x / y if y != 0 else 0
pres = add(5,10)
print(pres)
print(add(10,10))
print(div(10, 0))

15
20
0


In [ ]:
print((lambda x, y: x + y)(3, 7))

#### Map/Filter/Reduce - це вбудовані функції у Python, які дозволяють працювати з ітерабельними об'єктами. 
>**Map** - застосувати функцію до кожного об'єкта в колекції \
>**Filter** - фільтрація за допомогою зовнішньої функції \
>**Reduce** - Агрегаційна функція 


In [ ]:
# Map - map(function, iter_object) -> map_object -> some_object(map_object)

In [81]:
for i in range(0, 10):
    print(i ** 2)

0
1
4
9
16
25
36
49
64
81


In [84]:
int('15')

15

In [85]:
arr = ['1', '2', '3', '4', '5', '6', '7']

result = []
for i in arr:
    result.append(int(i))
print(result)

[1, 2, 3, 4, 5, 6, 7]


In [86]:
# lambda param,param2 : int(param)
sq = list(map(lambda x: int(x), arr))
print(sq)

[1, 2, 3, 4, 5, 6, 7]


In [ ]:
help(map)

In [6]:
# Map - примінити функцію для всіх 
# map(function, iterable object) - загальна структура
numbers = [i for i in range(0, 15)]

def square(x: int) -> int:
    return x ** 2

sq = map(square, numbers)
print(list(sq), sq)


# Або використати функцію 
sq1 = map(lambda x: x ** 2, numbers) # __iter__
print(list(sq1))

[0, 1, 4, 9, 16, 25, 36, 49, 64, 81, 100, 121, 144, 169, 196] <map object at 0x10f0ff820>
[0, 1, 4, 9, 16, 25, 36, 49, 64, 81, 100, 121, 144, 169, 196]


In [83]:
square(15)

225

In [ ]:
help(filter)

In [7]:
# filter - filter(function, iterable object)

even = tuple(filter(lambda x: x % 2 == 0, numbers))
print(even)

(0, 2, 4, 6, 8, 10, 12, 14)


In [ ]:
help(reduce)

In [1]:
from functools import reduce # потрібно імпортувати із built-in бібліотеки

numbers = [i for i in range(1, 16)]

product = reduce(lambda x, y: x+y, numbers)
print(product)  # 1 * 2 * 3 * 4 * 5
# 1 , 2 -> 2
# 2 , 3 -> 6
# etc...

120


In [15]:
numbers = [i for i in range(1, 5000)]

In [16]:
%%time
sum(number)

CPU times: user 6 µs, sys: 1 µs, total: 7 µs
Wall time: 9.3 µs


45

In [17]:
%%time
product = reduce(lambda x, y: x+y, numbers)
print(product)

12497500
CPU times: user 310 µs, sys: 4 µs, total: 314 µs
Wall time: 317 µs


In [10]:
sum(numbers)

120

In [22]:
# str 
string = ['1', '2', '3', '4', '5', '6', '7']

result = reduce(lambda x,y: x + y , map(lambda x: int(x), string))
print(result)

28


# завдання 
1. Використайте map() для перетворення списку зі рядок на список чисел: ['1', '2', '3', '4', '5'] → [1, 2, 3, 4, 5].
2. Використайте filter() для знаходження всіх парних чисел у списку: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10] → [2, 4, 6, 8, 10].
3. Використайте reduce() для обчислення добутку всіх елементів у списку: [1, 2, 3, 4, 5] → 120 (1 * 2 * 3 * 4 * 5).
4. Використайте map() для перетворення списку зі рядок на список квадратів цих чисел: [1, 2, 3, 4, 5] → [1, 4, 9, 16, 25].
5. Використайте filter() для знаходження всіх позитивне чисел у списку: [1, -2, 3, -4, 5, -6, 7, -8, 9, -10] → [1, 3, 5, 7, 9].
6. Використайте reduce() для обчислення суми всіх елементів у списку: [1, 2, 3, 4, 5] → 15 (1 + 2 + 3 + 4 + 5).
7. Використайте map() для перетворення списку зі рядок на список їхніх довжин: ['apple', 'banana', 'cherry'] → [5, 6, 6].
8. Використайте filter() для знаходження всіх елементів списку, які більше за середнє значення списку: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10] → [6, 7, 8, 9, 10].
9. Використайте reduce() для обчислення найбільшого елемента у списку: [1, 5, 3, 9, 2] → 9.
10. Використайте map() для перетворення списку зі рядок на список великих літер: ['apple', 'banana', 'cherry'] → ['APPLE', 'BANANA', 'CHERRY'].

In [ ]:
# Task 1
number = [str(i) for i  in range(0, 10)]

convert_list = map(lambda x: int(x), number)
print(list(convert_list))

In [23]:
# Task 2
number = [i for i in range(1, 11)]

even = tuple(filter(lambda x: x % 2 == 0, number))

print(number, even)

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10] (2, 4, 6, 8, 10)


In [24]:
# Task 3

number = [i for i in range(1, 11)]

result = reduce(lambda x,y: x * y, number)
print(result)

3628800


In [25]:
# Task 4

number = [i for i in range(1, 11)]

result = tuple(map(lambda x: x ** 2, number))
print(result)

(1, 4, 9, 16, 25, 36, 49, 64, 81, 100)


In [ ]:
# Task 5
arr = [-1,30, 0, 60, -5, -10, -100, 121]

result = list(filter(lambda x: x >= 0, arr))

print(result)

In [28]:
def dec1(x = lambda x: x ** 2):
    pass

In [4]:
# Task 6
arr = [i for i in range(1, 100)]

# first var
result = reduce(lambda x, y: x + y, arr)
print(result)

# second var
result = sum(i for i in arr if isinstance(i, (int, float)))
print(result) 

# thierd var
result = sum(arr)
print(result)

# 4-var
count_ = 0
for i in arr:
    count_ += i 
print(count_)



4950
4950
4950
4950


In [8]:
# Task 7 

# var 1 
arr = ['apple', 'banana', 'cherry']

result = list(map(lambda x: (x,len(x)), arr))

print(arr, result)


['apple', 'banana', 'cherry'] [('apple', 5), ('banana', 6), ('cherry', 6)]


In [12]:
# Task 8 

arr = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

# 1 var 
avg_ = sum(arr) / len(arr)

result = list(filter(lambda x: x > avg_, arr))
print(avg_, result)

# 2 var 
from statistics import mean 

avg_ = mean(arr) 
result = list(filter(lambda x: x > avg_, arr))
print(avg_, result)


5.5 [6, 7, 8, 9, 10]
5.5 [6, 7, 8, 9, 10]


In [14]:
# Task 9 

# var 1
arr = [1, 5, 3, 9, 2]
print(max(arr), arr)

# var 2
print(sorted(arr)[-1], arr)

# var 3 
result = reduce(lambda x, y: max(x, y), arr)
print(result, arr)


9 [1, 5, 3, 9, 2]
9 [1, 5, 3, 9, 2]
9 [1, 5, 3, 9, 2]


In [15]:
# Task 10 
arr = ['apple', 'banana', 'cherry']

result = list(map(lambda x: x.upper(), arr))
print(arr, result)

['apple', 'banana', 'cherry'] ['APPLE', 'BANANA', 'CHERRY']


#### try/except - опрацювання помилок, запуск альтернативного коду в залежності від помилки 

In [29]:
1 / 0

ZeroDivisionError: division by zero

In [30]:
float('*')

ValueError: could not convert string to float: '*'

In [42]:
try:
    number = float(input('Input value: '))
   # print(number / 0)
   # print(str(number)[10])
# except ValueError: 
#     print('Wrong value input ')
except ZeroDivisionError:
    print('ZeroDivisionError')
except Exception as e:
    print(f'Wrong data, you error is {e} {e.with_traceback}')
finally:
    print('All okey')

Wrong data, you error is could not convert string to float: '*' <built-in method with_traceback of ValueError object at 0x10f147ec0>


In [44]:
try_numbers = 1
while try_numbers < 2:
    try:
        number = float(input('Input value: '))
    # print(number / 0)
    # print(str(number)[10])
    # except ValueError: 
    #     print('Wrong value input ')
    except ZeroDivisionError:
        print('ZeroDivisionError')
    except Exception as e:
        print(f'Wrong data, you error is {e} {e.with_traceback}')
    finally:
        print(f'Try number: {try_numbers}')
        try_numbers += 1

Wrong data, you error is could not convert string to float: '*' <built-in method with_traceback of ValueError object at 0x10f5884f0>
Try number: 1


In [1]:
number = [ i for i in range(0, 10)]
print(number[10])

IndexError: list index out of range

In [ ]:
# Task: try/except
# 1. Написати калькультор, який при діленні на нуль повертає текст помилки, опрацювати інші можливі помилки також 
# 2. Написати функцію, яка при вводі юзера не числа, буде повторювати дію, до тих пір поки user - не введе позитивне число 

In [18]:
import random

random.randint(1, 150)

118

In [ ]:
# Read file tasks
# 1. Створити файл та записати в нього словних: {'UK': 'London', 'Germany': 'Berlin', 'Ukraine': 'Kyiv'}
# Формат запису: Country: name_of_contry. Capital is: name_of_capital

# 2. Прочитати файл Japan, та в нього добавити інформацію про державний устрій(Кон монархія). Зробити це за допомогою context manager

# 3. Прочитати файл Japan та записати його зміст в інший файл Japan_Backup в назву файлу добавити поточну дату, зробити це за допомогою context_manager

# 4. Створити файли де назвою буде алф від A-Z, записати в них будь-яке число(бібліотека random), зробити за допомогою context_manager


In [21]:
# Task 1
capitals = {'UK': 'London', 'Germany': 'Berlin', 'Ukraine': 'Kyiv'}
 
with open('capitals.txt', 'w') as file:
    for country, capital in capitals.items():
        file.writelines(f'Country: {country}. Capital is: {capital}\n')

In [22]:
# Task 2

with open('Japan.txt', 'a') as file:
    file.write('\nустрій(Кон монархія)\n')

In [34]:
# Task 3
import datetime 

now = datetime.datetime.today()

with open('Japan.txt', 'r') as file, open(f'JapanBackup{now}.txt', 'w') as back_file:
    for i in file.readlines():
        back_file.writelines(i)

In [64]:
import requests


ModuleNotFoundError: No module named 'requests'

In [38]:
# Task 4
import random as rn    

def generate_value(n: int = 1, m: int = 100) -> int:
    return str(rn.randint(n, m)) if m > n else '1'


def generate_alp(start_number: int = 65, end_number: int = 90) -> str:
    return ''.join(chr(i) for i in range(start_number, end_number + 1 )) if end_number > start_number else None 


for item in generate_alp():
    with open(f'{item}.txt', 'w') as file:
        file.write(generate_value())

In [61]:
ord('А'), ord('Ґ')

(1040, 1168)

In [45]:
generate_value(n = 100, m = 110)

'103'